pkg/statedb/reflector/k8s_test.go

```diff
@@ -0,0 +1,260 @@
+		)
+
+		return reflector.KubernetesConfig[*v1.Pod]{
+			BufferSize:     batchSize,
+			BufferWaitTime: time.Hour, // We want to only emit on full buffer
+			ListerWatcher:  lw,
+			Table:          table,
+		}
+	}
+
+	// Insert a new pod and wait for table to update.
+	expectedPod := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{APIVersion: "v1", Kind: "Pod"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "pod1", Namespace: "test1",
+			Labels: map[string]string{
+				"foo": "bar",
+			},
+		},
+	}
+
+	h := hive.New(
+		statedb.Cell,
+		job.Cell,
+		client.FakeClientCell,
+
+		cell.Invoke(func(cs client.Clientset) {
+			// Create the initial pods before anything starts. The fake client
+			// doesn't know how to "Watch" from a specific ResourceVersion so we
+			// need to make sure that we've added everything before Watch() gets called.
+			for j := 0; j < batchSize; j++ {
+				expectedPod.ObjectMeta.Name = fmt.Sprintf("pod%d", j)
+				expectedPod.ObjectMeta.ResourceVersion = "1"
+				expectedPod.ObjectMeta.Generation = 1
+				expectedPod.UID = types.UID(expectedPod.Name)
+				_, err = cs.CoreV1().Pods("test1").Create(
+					context.Background(), expectedPod.DeepCopy(), metav1.CreateOptions{})
+				require.NoError(b, err, "Create pod")
+			}
+		}),
+
+		cell.Module("test", "Test",
+			cell.ProvidePrivate(sourceConfig),
+			reflector.KubernetesCell[*v1.Pod](),
+
+			cell.Invoke(
+				func(db *statedb.DB) error {
+					return db.RegisterTable(table)
+				}),
+
+			cell.Invoke(func(p_ params) { p = p_ }),
+		),
+	)
+
+	require.NoError(b, h.Start(context.TODO()))
+
+	// Wait for pods to be created
+	iter, watch := table.All(p.DB.ReadTxn())
+	for {
+		objs := statedb.Collect[*v1.Pod](iter)
+		if len(objs) == batchSize {
+			break
+		}
+		<-watch
+		iter, watch = table.All(p.DB.ReadTxn())
+	}
+
+	// Benchmark updating a batch of objects and checking that the table
+	// has updated. Actual throughput is 'batchSize' * the number of iterations
+	// in a second. On my i5 laptop I'm getting ~120k/s with batchSize of 100.
+
+	b.Logf("batchSize=%d, multiple the ops/sec with it for per-object throughput", batchSize)
+
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		expectedPod.ResourceVersion = fmt.Sprintf("%d", i+2)
+		expectedPod.Generation = int64(i + 1)
+		expectedPod.Labels["i"] = fmt.Sprintf("%d", i)
+		for j := 0; j < batchSize; j++ {
+			expectedPod.Name = fmt.Sprintf("pod%d", j)
+			expectedPod.UID = types.UID(expectedPod.Name)
+			_, err = p.Clientset.CoreV1().Pods("test1").Update(
+				context.Background(), expectedPod.DeepCopy(), metav1.UpdateOptions{})
+			require.NoError(b, err, "Update pod")
+		}
+
+		// Wait until the batch has committed.
+		<-watch
+
+		// Check that all objects are from the new round
+		iter, watch1 := table.All(p.DB.ReadTxn())
+		watch = watch1
+		objs := statedb.Collect[*v1.Pod](iter)
+		require.Len(b, objs, batchSize)
+		for _, obj := range objs {
+			require.Equal(b, obj.Labels["i"], fmt.Sprintf("%d", i))
+		}
+
+	}
```

Comment:
We can output the throughput as a metric and make it part of the standard benchmark output (and thus comparable with benchstat too) using:

```
	b.StopTimer()
	b.ReportMetric(float64(batchSize)*float64(iterations)/b.Elapsed().Seconds(), "ops/sec")
```

at the end of the `for` loop.
`iterations` here is incremented once for each loop iteration.

Reply:
Ah excellent, TIL.